README
Catalyst-Engine-Stomp
=====================

This is a Catalyst engine which allows you to build a message-driven
application with STOMP. 

Controllers indicate queues, and message types indicate actions. A
controller base class is provided to handle YAML-serialized messages.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

     Moose
     Net::Stomp
     YAML::XS
     Data::Serializer
     Catalyst::Engine::Embeddable
     Catalyst::Runtime 5.80003
     namespace::autoclean
     
The supplied workers script requires MooseX::Workers, and you'll need
the relevant modules to support Data::Serializer if you choose
something other than YAML as a message format. 

COPYRIGHT AND LICENCE

Copyright (C) 2009 Venda Ltd
Copyright (C) 2011 Paul Mooney
Copyright (C) 2001 Gianni Ceccarelli

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Catalyst::Engine::Stomp.

0.16  Fri  6 May 16:00 2011
    - Thanks to chiselwright for flushing the logs when an error is thrown
      and not using 100% CPU when unable to connect.

0.15  Mon  7 Feb 14:00 2011
    - Thanks to dakkar for:
    - Config option tries_per_server is now optional
    - Only join a queue once even if specified twice!
    - Code tidy

0.14  Tue  8 Jue 17:00
    - More than one stomp server can be specified so a client can
      failover to the next server in the list

0.13  Wed 10 Mar 16:45
    - Added ability to send signal USR1 to the Catalyst app so no more
      requests are handled and it exists.

0.12  Tue Dec  7 17:45
    - The raw request (in say YAML) is kept in $c->stash->{raw_request}
      which is useful for debugging/logging

0.11  Fri Dec  4 16:30:00 2009
    - Really really fix the tests this time

0.10  Tue Dec  1 16:00:00 2009
    - Fixed tests _only_ to allow for cleaner install

0.09  Fri Nov 27 17:00:00 2009
    - If a single error is exists and it is blessed, that is serialized
      and sent back to the user. Before it was stringyfied.

0.08  Thu Oct 29 17:00:00 2009
    - Added ability to convert message to UTF-8 on output if flag set
    - Added test to demonstrate Throwable error objects not being
      passed up the stack correctly

0.07  Thu Aug 27 22:37:46 2009
    - Ability to set the message key used to dispatch
      to a message-driven Controller action (Jason Tang)
    - Ability to add additional header key/value pairs to the
      subscribe message sent to the message broker. (Jason Tang)
    - Additional documentation (Jason Tang)

0.06  Mon Jul 13 22:22:22 2009
    - Make the test-app harness available under CatalystX::
    - Clean up use of deprecated dispatcher methods.
    - Add an error message in the Catalyst log when we
      receive an unserializeable message.
    - Fix the test application to always run with the
      correct library paths.
    - Fix bug causing alternate serializers to not be found
      when using configuration from a config file.

0.05  Sun Jun 07 01:38:46 2009
    - Cleanups, tests and documentation additions.
    - Add namespace:autoclean
    - Handle 'no type' messages better, attempt to reply saying so.

0.04  Mon Jun 01 09:21:00 2009
    - Remove dependency on Catalyst::Action::REST, use
      Data::Serializer directly instead.

0.03  Thu May 21 22:57:00 2009
	- Add Alien::ActiveMQ for real testing.

0.02  Sun May 17 00:00:00 2009
	- Rename TestApp

0.01  Fri May 15 21:39:30 2009
	- Initial CPAN Release



t/15_utf8.pl
use strict;
use warnings;
use Test::More;

=pod

Tests which expect a STOMP server like ActiveMQ to exist on
localhost:61613, which is what you get if you just get the ActiveMQ
distro and run its out-of-the-box config.

If the Load() function in YAML::XS is given a byte that can be thought
of as the first of a multibyte character (UTF-8) and it isn't, it can
explode.

=cut

use Net::Stomp;
use YAML::XS qw/ Dump Load /;
use Data::Dumper;
use Encode;

use FindBin;
use lib "$FindBin::Bin/lib";
use lib "$FindBin::Bin/../lib";
use TestServer;

binmode(STDOUT, ":utf8"); # Set you terminal to show UTF8 !!!!!!!!!!

my $stomp = start_server();

plan tests => 13;

my $frame = $stomp->connect();
ok($frame, 'connect to MQ server ok');

my $reply_to = sprintf '%s:1', $frame->headers->{session};
ok($frame->headers->{session}, 'got a session');
ok(length $reply_to > 2, 'valid-looking reply_to queue');

ok($stomp->subscribe( { destination => '/temp-queue/reply' } ), 'subscribe to temp queue');

# \x{eb} is ë, or "LATIN SMALL LETTER E WITH DIAERESIS".
# \x{eb} is the codepoint, 235, perl uses internally for the character
# If you have ë in a simple perl string, switch the utf-8 flag on, Dumper() will display \x{eb}
# In octets, i.e. proper UTF-8, it comes out as \xc3 \xab,
# So we want two bytes back, not one
# The Load() function will convert it back into Perl internal representation of \x{eb}
#

my $text_string  = "Maria van Bourgondiëlaan";
Encode::_utf8_on($text_string); # this is now how we get things from XML::Compile which deals with XML in utf-8 encoding

my $message = {
	       payload => { string => $text_string }, # was send_string...
	       reply_to => $reply_to,
	       type => 'testutf8',
	      };
my $text = Dump($message);
ok($text, 'compose message, sending : ' . Dumper($message) );

$stomp->send( { destination => '/queue/testcontroller', body => $text } );

ok(1, "sent");

my $reply_frame = $stomp->receive_frame();
ok($reply_frame, 'got a reply');
ok($reply_frame->headers->{destination} eq "/remote-temp-queue/$reply_to", 'came to correct temp queue');
ok($reply_frame->body, 'has a body');

my $response = Load($reply_frame->body);

ok($response, 'YAML response ok');
ok($response->{type}   eq 'testutf8_response', 'correct type');

# Without utf-8 encoding, the new string comes back as "Maria van Bourgondi\x{fffd}laan"
# With    utf-8 encoding, the new string comes back as "Maria van Bourgondi\x{eb}laan"
my $new_string = $response->{struct}->[1]->{new_string};

ok($new_string eq $text_string,
   "new string is the same as our text string : \n" .
   Dumper( $new_string ) . " == \n" . Dumper($text_string) .
   "OR\n".
   "$new_string eq $text_string\n".
   "utf8 for new_string  : " . Encode::is_utf8($new_string)  . "\n" .
   "utf8 for text_string : " . Encode::is_utf8($text_string) . "\n"  );


$stomp->disconnect;
ok(!$stomp->socket->connected, 'disconnected');


t/16_old_config.pl
use strict;
use warnings;
use Test::More;

# Tests which expect a STOMP server like ActiveMQ to exist on
# localhost:61613, which is what you get if you just get the ActiveMQ
# distro and run its out-of-the-box config.

use Net::Stomp;
use YAML::XS qw/ Dump Load /;
use Data::Dumper;

use FindBin;
use lib "$FindBin::Bin/lib";
use TestServer;

my $stomp = start_server("$FindBin::Bin/lib/StompTestApp/stomptestapp_old.yml");

plan tests => 11;

my $frame = $stomp->connect();
ok($frame, 'connect to MQ server ok');

my $reply_to = sprintf '%s:1', $frame->headers->{session};
ok($frame->headers->{session}, 'got a session');
ok(length $reply_to > 2, 'valid-looking reply_to queue');

ok($stomp->subscribe( { destination => '/temp-queue/reply' } ), 'subscribe to temp queue');

my $message = {
	       payload => { foo => 1, bar => 2 },
	       reply_to => $reply_to,
	       type => 'testaction',
	      };
my $text = Dump($message);
ok($text, 'compose message');

$stomp->send( { destination => '/queue/testcontroller', body => $text } );

my $reply_frame = $stomp->receive_frame();
ok($reply_frame, 'got a reply');
ok($reply_frame->headers->{destination} eq "/remote-temp-queue/$reply_to", 'came to correct temp queue');
ok($reply_frame->body, 'has a body');

my $response = Load($reply_frame->body);
ok($response, 'YAML response ok');
ok($response->{type} eq 'testaction_response', 'correct type');

$stomp->disconnect;
ok(!$stomp->socket->connected, 'disconnected');



t/lib/TestServer.pm
package StompRole;
use Moose::Role;
use namespace::autoclean;

after 'disconnect' => sub {
    delete shift->{___activemq};
};

package TestServer;
use strict;
use warnings;
use Alien::ActiveMQ;
use Test::More;
use Exporter qw/import/;
use FindBin;

our $ACTIVEMQ_VERSION = '5.2.0';

our @EXPORT = qw/ start_server /;

sub start_server {
    my ($mq, $stomp);
    eval {
        $stomp = Net::Stomp->new( { hostname => 'localhost', port => 61613 } );
    };
    if ($@) {

        unless (Alien::ActiveMQ->is_version_installed($ACTIVEMQ_VERSION)) {
            plan 'skip_all' => 'No ActiveMQ server installed by Alien::ActiveMQ, try running the "install-activemq" command';
            exit;
        }

        $mq = Alien::ActiveMQ->run_server($ACTIVEMQ_VERSION);

        eval {
            $stomp = Net::Stomp->new( { hostname => 'localhost', port => 61613 } );
        };
        if ($@) {
            plan 'skip_all' => 'No ActiveMQ server listening on 61613: ' . $@;
            exit;
        }
    }

    $SIG{CHLD} = 'IGNORE';
    unless (fork()) {
        my $libs = join(' ', map { "-I$_" } @INC);
        $ENV{CATALYST_CONFIG}=shift if @_;
        system("$^X $libs $FindBin::Bin/script/stomptestapp_stomp.pl --oneshot");

        # Let our tests complete - we need to sleep here otherwise we get a timing issue
        # problem which I don't fully understand. Without it, sometimes the test works,
        # somethimes it doesn't!
        sleep 2;

        exit 0;
    }
    print STDERR "server started, waiting for spinup...";
    sleep 20;

    $stomp->{___activemq} = $mq if $mq;
    StompRole->meta->apply($stomp);
    return $stomp;
}



Makefile.PL
use inc::Module::Install 0.91;
use strict;
use warnings;
use 5.008006;

BEGIN {
    if ($Module::Install::AUTHOR) {
        require Module::Install::AuthorRequires;
    }
}

license  'perl';

name     'Catalyst-Engine-Stomp';
all_from 'lib/Catalyst/Engine/Stomp.pm';

requires 'Catalyst::Engine::Embeddable'  => '0.0.1';
requires 'Catalyst::Runtime'  => '5.80004';
requires 'Moose' => undef;
requires 'MooseX::Types' => undef;
requires 'MooseX::Workers' => '0.05';
requires 'Net::Stomp' => '0.34';
requires 'YAML::XS' => '0.32';
requires 'YAML' => undef;
requires 'Data::Serializer' => '0.49';
requires 'namespace::autoclean' => '0.05';

test_requires 'Alien::ActiveMQ' => '0.00003';
test_requires 'Catalyst::Plugin::ConfigLoader' => undef;

author_requires 'Test::NoTabs';
author_requires 'Test::Pod';
author_requires 'Test::Pod::Coverage';
author_requires 'Pod::Coverage';

if ($Module::Install::AUTHOR) {
    darwin_check_no_resource_forks();
}

resources
    'repository' => 'git://github.com/chrisa/catalyst-engine-stomp.git',
    'license' => 'http://dev.perl.org/licenses/';

install_script glob('script/*.pl');
auto_install;
WriteAll;

sub darwin_check_no_resource_forks {
    if ($^O eq 'darwin') {
        my $osx_ver = `/usr/bin/sw_vers -productVersion`;
        chomp $osx_ver;

        # TAR on 10.4 wants COPY_EXTENDED_ATTRIBUTES_DISABLE
        # On 10.5 (Leopard) it wants COPYFILE_DISABLE
        my $attr = $osx_ver =~ /^10.5/ ? 'COPYFILE_DISABLE' : 'COPY_EXTENDED_ATTRIBUTES_DISABLE';

        makemaker_args(dist => { PREOP => qq{\@if [ "\$\$$attr" != "true" ]; then}.
                                          qq{ echo "You must set the ENV variable $attr to true,"; }.
                                          ' echo "to avoid getting resource forks in your dist."; exit 255; fi' });
        }
}



lib/Catalyst/Controller/MessageDriven.pm
package Catalyst::Controller::MessageDriven;
use Moose;
use Data::Serializer;
use Moose::Util::TypeConstraints;
use MooseX::Types::Moose qw/Str/;
use namespace::autoclean;

BEGIN { extends 'Catalyst::Controller' }

=head1 NAME

Catalyst::Controller::MessageDriven

=head1 SYNOPSIS

  package MyApp::Controller::Queue;
  use Moose;
  BEGIN { extends 'Catalyst::Controller::MessageDriven' }

  sub some_action : Local {
      my ($self, $c, $message) = @_;

      # Handle message

      # Reply with a minimal response message
      my $response = { type => 'testaction_response' };
      $c->stash->{response} = $response;
  }

=head1 DESCRIPTION

A Catalyst controller base class for use with Catalyst::Engine::Stomp,
which handles YAML-serialized messages. A top-level "type" key in the
YAML determines the action dispatched to.

=head1 METHODS

=head2 begin

Deserializes the request into C<< $c->stash->{request} >>

=head2 default

Dispatches to method named by the key C<< $c->stash->{request}->{type} >>

=head2 end

Serializes the response from C<< $c->stash->{response} >>

=head1 CONFIGURATION

In the configuration file add the following to set the value for a parameter

  <MessageDriven>
    type_key foo
  </MessageDriven>

=head2 type_key

The hash key the module will try to pull out the received message to call
within the controller. This defaults to 'type'.

=head2 serializer

The serializer used to serialiser/deserialise. See Data::Serializer to see
what is available. Defaults to YAML. JSON is anotther that is available.


=cut

class_type 'Data::Serializer';
my $serializer_t = subtype 'Data::Serializer', where { 1 };
coerce $serializer_t, from 'Str',
    via { Data::Serializer->new( serializer => $_ ) };

has serializer => (
    isa => $serializer_t, is => 'ro', required => 1,
    default => 'YAML', coerce => 1,
);

has type_key => (
    is => 'ro', required =>1,
    default => 'type',
);


sub begin : Private {
    my ($self, $c) = @_;

    # Deserialize the request message
    my $message;
    my $s = $self->serializer;
    eval {
        my $body = $c->request->body;
        open my $IN, "$body" or die "can't open temp file $body";
        my $raw_request = do { local $/; <$IN> };
        $c->stash->{raw_request} = $raw_request;
        $message = $s->raw_deserialize($raw_request);
    };
    if ($@) {
        # can't reply - reply_to is embedded in the message
        $c->error("exception in deserialize: $@");
    }
    else {
        $c->stash->{request} = $message;
    }
}

sub _errors_to_response {
    my ($self, $c) = @_;

    if ( scalar(@{$c->error}) == 1 && ref($c->error->[0]) ) {
        # A single object exists as an error, throw that back as is
        $c->log->error('Exception thrown: ' . $c->error->[0]);
        return $c->error->[0];
    }
    else {
        $c->log->error($_) for @{$c->error}; # Log errors in Catalyst
        my $error = join "\n", @{$c->error}; # Stringyfy them
        return { status => 'ERROR', error => $error};
    }

    return;
}

sub end : Private {
    my ($self, $c) = @_;

    # Engine will send our reply based on the value of this header.
    $c->response->headers->header( 'X-Reply-Address' => $c->stash->{request}->{reply_to} );

    # The wire response
    my $output;

    # Load a serializer
    my $s = $self->serializer;

    # Custom error handler - steal errors from catalyst and dump them into
    # the stash, to get them serialized out as the reply.
    if (scalar @{$c->error}) {
        $c->stash->{response} = $self->_errors_to_response($c);
        $c->clear_errors;
        $c->response->status(400);
    }

    # Serialize the response
    eval {
        $output = $s->raw_serialize( $c->stash->{response} );
    };
    if ($@) {
         my $error = "exception in serialize: $@";
         $c->stash->{response} = { status => 'ERROR', error => $error };
         $output = $s->serialize( $c->stash->{response} );
        $c->response->status(400);
    }

    $c->response->output( $output );
}

sub default : Private {
    my ($self, $c) = @_;

    # Forward the request to the appropriate action, based on the
    # message type.
    my $action = $c->stash->{request}->{ $self->type_key };
    if (defined $action) {
        $c->forward($action, [$c->stash->{request}]);
    }
    else {
         $c->error('no message type specified');
    }
}

__PACKAGE__->meta->make_immutable;

=head1 AUTHOR AND CONTRIBUTORS

See information in L<Catalyst::Engine::Stomp>

=head1 LICENCE AND COPYRIGHT

Copyright (C) 2009 Venda Ltd

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut



lib/Catalyst/Engine/Stomp.pm
package Catalyst::Engine::Stomp;
use Moose;
use List::MoreUtils qw/ uniq /;
use HTTP::Request;
use Net::Stomp;
use MooseX::Types::Moose qw/Str Int HashRef/;
use namespace::autoclean;
use Encode;

extends 'Catalyst::Engine::Embeddable';

our $VERSION = '0.16';

has connection => (is => 'rw', isa => 'Net::Stomp');
has conn_desc => (is => 'rw', isa => Str);

=head1 NAME

Catalyst::Engine::Stomp - write message handling apps with Catalyst.

=head1 SYNOPSIS

  # In a server script:

  BEGIN {
    $ENV{CATALYST_ENGINE} = 'Stomp';
    require Catalyst::Engine::Stomp;
  }

  MyApp->config(
    Engine::Stomp' = {
       tries_per_server => 3,
      'servers' => [
       {
         'hostname' => 'localhost',
         'port' => '61613'
       },
       {
         'hostname' => 'stomp.yourmachine.com',
         'port' => '61613'
       },
       utf8             => 1,
       subscribe_header => {
         transformation       => 'jms-to-json',
       }
    },
  );
  MyApp->run();

  # In a controller, or controller base class:
  use base qw/ Catalyst::Controller::MessageDriven /;

  # then create actions, which map as message types
  sub testaction : Local {
      my ($self, $c) = @_;

      # Reply with a minimal response message
      my $response = { type => 'testaction_response' };
      $c->stash->{response} = $response;
  }

=head1 DESCRIPTION

Write a Catalyst app connected to a Stomp messagebroker, not HTTP. You
need a controller that understands messaging, as well as this engine.

This is single-threaded and single process - you need to run multiple
instances of this engine to get concurrency, and configure your broker
to load-balance across multiple consumers of the same queue.

Controllers are mapped to Stomp queues, and a controller base class is
provided, Catalyst::Controller::MessageDriven, which implements
YAML-serialized messages, mapping a top-level YAML "type" key to
the action.

=head1 FAILOVER

You can specify one or more servers in a list for the apps config.
This enables fail over if an error occurs, like the broker or network
connection disappears.

It will try to use a server a set number of times, as determined by
tries_per_server in the config before failing on to the next server
in the list. It cycle through each server in turn, going back to the
start of the list if need be.

=head1 UTF-8

By default STOMP messages are assumed to be in UTF-8. This module can
automatically convert a Perl string into a UTF-8 set of octets to be
sent over the wire instead. This is a Good Thing, especially if you 
use the function Load() from the package YAML::XS to un-serialize
it in your client - it assumes it is in UTF-8.

If you do want this behaviour, set 'utf8' to '1' in your config.

=head1 Simplified configuration

Instead of using the complete config layout as shown in the synopsis,
you can

=over 4

=item *

not specify a C<tries_per_server> (defaults to 1)

=item *

specify a single server:

  server => { hostname => $host, port => $port }

=item *

use the old-style (pre 0.14) config having C<hostname> and C<port>
directly in the C<Engine::Stomp> block, without a C<server> key in
between.

=back

=cut

=head1 METHODS

=head2 _see_ya

Set to run when signal USR1 is received. Simply sets the stop flag.

=cut

my $stop = 0;

sub _see_ya {
    $stop = 1;
    delete $SIG{'USR1'};
}


=head2 run

App entry point. Starts a loop listening for messages.

If the stop flag is set (see _see_ya above) then no more requests are processed.
Keep in mind this is a blocking server and it will wait for a STOMP message forever.
Only after handling a request does it check the flag.

=cut

sub run {
    my ($self, $app, $oneshot) = @_;

    $SIG{'USR1'} = \&_see_ya;

    die 'No Engine::Stomp configuration found'
        unless ref $app->config->{'Engine::Stomp'} eq 'HASH';

    my @queues = uniq grep { length $_ }
                 map  { $app->controller($_)->action_namespace } $app->controllers;

    # connect up
    my $config = $app->config->{'Engine::Stomp'};
    my $index  = 0;

    # munge the configuration to make it easier to write
    $config->{tries_per_server} ||= 1;
    $config->{connect_retry_delay} ||= 15;
    if (! $config->{servers} ) {
        $config->{servers} = [ {
            hostname => (delete $config->{hostname}),
            port => (delete $config->{port}),
        } ];
    }
    elsif (ref $config->{servers} eq 'HASH') {
        $config->{servers} = [ $config->{servers} ];
    }

    QUITLOOP:
    while (1) {
        # Go to next server in list
        my %template = %{ $config->{servers}->[$index] };
        $config->{hostname} = $template{hostname};
        $config->{port}     = $template{port};

        ++$index;

        if ($index >= (scalar( @{$config->{servers}} ))) {
            $index = 0; # go back to first server in list
        }

        my $tries = 0;

        while ($tries < $config->{tries_per_server}) {
            ++$tries;
    
            eval {
                my $subscribe_headers = $template{subscribe_headers} || {};
                die("subscribe_headers config for Engine::Stomp must be a hashref!\n")
                    if (ref($subscribe_headers) ne 'HASH');

                $app->log->info("Connecting to STOMP Q at " . $template{hostname}.':'.$template{port});

                $self->connection(Net::Stomp->new(\%template));
                $self->connection->connect();
                $self->conn_desc($template{hostname}.':'.$template{port});

                # subscribe, with client ack.
                foreach my $queue (@queues) {
                    my $queue_name = "/queue/$queue";
                    $self->connection->subscribe({
                        %$subscribe_headers,
                        destination => $queue_name,
                        ack         => 'client',
                    });
                }

                # Since we might block for some time, lets flush the log messages
                $app->log->_flush() if $app->log->can('_flush');

                # enter loop...
                while (1) {
                    my $frame = $self->connection->receive_frame(); # block
                    $self->handle_stomp_frame($app, $frame);
            
                    if ( $ENV{ENGINE_ONESHOT} || $stop ){
                        # Perl does not like 'last QUITLOOP' inside an eval, hence we die and do it
                        die "QUITLOOP\n";
                    }
                }
            };

            if (my $err=$@) {
                # although it looks like a lot of pointless flush()ing we need
                # to make sure the user(s) can see any new messages; we
                # sometimes die before we flush() in the loop above

                if ($err eq "QUITLOOP\n") {
                    last QUITLOOP;
                }
                else {
                    $app->log->error(" Problem dealing with STOMP : $err");
                    $app->log->_flush() if $app->log->can('_flush');
                }

                # don't loop continuously if we can't connect; take a break;
                # give the service a chance to come back
                if ($err =~ m{Connection refused}) {
                    $app->log->info(
                          'Unable to connect to '
                        . $template{hostname}.':'.$template{port}
                        . '; sleeping before next retry'
                    );
                    $app->log->_flush() if $app->log->can('_flush');
                    sleep $config->{connect_retry_delay};
                }
            }
        }
    }
}

=head2 prepare_request

Overridden to add the source broker to the request, in place of the
client IP address.

=cut

sub prepare_request {
    my ($self, $c, $req, $res_ref) = @_;
    shift @_;
    $self->next::method(@_);
    $c->req->address($self->conn_desc);
}

=head2 finalize_headers

Overridden to dump out any errors encountered, since you won't get a #'
"debugging" message as for HTTP.

=cut

sub finalize_headers {
    my ($self, $c) = @_;
    my $error = join "\n", @{$c->error};
    if ($error) {
        $c->log->debug($error);
    }
    return $self->next::method($c);
}

=head2 handle_stomp_frame

Dispatch according to Stomp frame type.

=cut

sub handle_stomp_frame {
    my ($self, $app, $frame) = @_;

    my $command = $frame->command();
    if ($command eq 'MESSAGE') {
        $self->handle_stomp_message($app, $frame);
    }
    elsif ($command eq 'ERROR') {
        $self->handle_stomp_error($app, $frame);
    }
    else {
        $app->log->debug("Got unknown Stomp command: $command");
    }
}

=head2 handle_stomp_message

Dispatch a Stomp message into the Catalyst app.

=cut

sub handle_stomp_message {
    my ($self, $app, $frame) = @_;

    # queue -> controller
    my $queue = $frame->headers->{destination};
    my ($controller) = $queue =~ m|^/queue/(.*)$|;

    # set up request
    my $config = $app->config->{'Engine::Stomp'};
    my $url = 'stomp://'.$config->{hostname}.':'.$config->{port}.'/'.$controller;
    my $req = HTTP::Request->new(POST => $url);
    $req->content($frame->body);
    $req->content_length(length $frame->body);

    # dispatch
    my $response;
    $app->handle_request($req, \$response);

    # reply, if header set
    if (my $reply_to = $response->headers->header('X-Reply-Address')) {
        my $reply_queue = '/remote-temp-queue/' . $reply_to;
        my $content     = $response->content;

        if ($config->{utf8}) {
            $content = encode("utf8", $response->content); # create octets
        }

        $self->connection->send({ destination => $reply_queue, body => $content });
    }

    # ack the message off the queue now we've replied / processed
    $self->connection->ack( { frame => $frame } );
}

=head2 handle_stomp_error

Log any Stomp error frames we receive.

=cut

sub handle_stomp_error {
    my ($self, $app, $frame) = @_;

    my $error = $frame->headers->{message};
    $app->log->debug("Got Stomp error: $error");
}

__PACKAGE__->meta->make_immutable;

=head1 CONFIGURATION

=head2 subscribe_header

Add additional header key/value pairs to the subscribe message sent to the
message broker.

=cut

=head1 DEVELOPMENT

The source to Catalyst::Engine::Stomp is in github:

  http://github.com/pmooney/catalyst-engine-stomp

=head1 AUTHOR

Chris Andrews C<< <chris@nodnol.org> >>

=head1 CONTRIBUTORS

Tomas Doran (t0m) C<< <bobtfish@bobtfish.net> >>

Jason Tang

Paul Mooney

=head1 LICENCE AND COPYRIGHT

Copyright (C) 2009 Venda Ltd

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut

